### DISCUSSION TITLE
Forwarding of Application Layer APP - UdpSink in OMNeT++

### DISCUSSION THREAD

**Discussion by zhangteng0526** [omnetpp/omnetpp#983] [Q&A] [answered]
In the aodv example of OMNeT++, I set the application layer app of a node to UdpSink, which is specifically used to receive external messages. But I want it to implement forwarding function after receiving the message. Forward the message to the application layer app of another node. This app is also UdpSink, what should I do. I hope to receive your help and would like to express my gratitude here

I am trying to use the method of socket ->sendTo. But I don't know how to write the IP address of the node I want to receive as L3Address (assuming the IP address of the node is known to be 145.236.0.1). So, I hope to receive your help

**Comment by rhornig** [comment, marked as answer]
(Just nitpicking: AODV is an INET example, not a base OMNeT++ example).

You should write your own UDP application module for this. Use the `UdpEchoApp` as an example. The only difference is that you would forward the packet to a different destination and not back to the sender. If you add a `destinationAddress` parameter `L3AddressResolver().resolve(destinationAddress)` would resolve a string based IP address (or module name) to proper `L3Address` that can be used by the `sendTo` call. (see also the `UspBasicApp` for examples of sending to a specific address)

**Reply by zhangteng0526** [reply]
> (Just nitpicking: AODV is an INET example, not a base OMNeT++ example).
> 
> You should write your own UDP application module for this. Use the `UdpEchoApp` as an example. The only difference is that you would forward the packet to a different destination and not back to the sender. If you add a `destinationAddress` parameter `L3AddressResolver().resolve(destinationAddress)` would resolve a string based IP address (or module name) to proper `L3Address` that can be used by the `sendTo` call. (see also the `UspBasicApp` for examples of sending to a specific address)

Thank you, indeed my negligence (or not rigorous), AODV is an example of INET. So if I have some questions about the INET library in the future, can I still ask or discuss them here? Regarding your answer (create your own Udp application), Can I modify it in UdpEchoAPP, and will return to the developer to modify it to forward to a specific target (here you can refer to UdpBasicApp)？？

**Reply by rhornig** [reply]
As for the place for INET related discussions I'm quite torn and undecided. On one hand, INET has its own discussion board where it is much easier to reference code snippets, issues etc. On the other hand we would like to see the whole community on the OMNeT++ mailing list to move here and splitting the discussion boards and the community attention is not optimal. Sadly, I cannot easily move discussions between the `omnetpp` and `inet-framework` organization so that is also an impedance mismatch. I would say, let's gather and discuss everything here for the moment and lets hope that github will allow issue/discussion migration between organizations in the future.

Regarding modifying the `UdpEchoApp`, you can do that, but it will most likely not accepted as a pull request (if you mean that) as an EchoApp echos the packets back as its name suggest. If it would forward/route it to somewhere else, it would be called `UdpPacketForwarderApp` or somethng similar. There are a myriad of ways to enhance modify these apps and not everything lends itself to easily allow to parameterize it. How do you decide where to send the packet. Should it be a fixed address? or the next address on my subnetwork, or choose randomnly from a destination addrress list? or round robin? Should the packet length remain, or should it change? If yes, how? Do I need to forward the packet immediately or is there a processing delay? If there is, are there a queue in there? And so on and on. If you don't want to create new C++ code for your app, you can also use the `UdpSocketIo` module and create an app by wiring together various modules from the queueing framework. As an example, check the `UdpApp` compound module and the `UdpRequestResponse` module. Maybe they are better starting points that the `UdpEchoApp` as they are more composable.
